acme/acme/client.py

```diff
@@ -48,6 +48,8 @@ class Client(object):  # pylint: disable=too-many-instance-attributes
 
     :ivar messages.Directory directory:
     :ivar key: `.JWK` (private)
+    :ivar account: `.Account` (private)
+    :ivar acme_version: `int` (private)
```

Comment:
> Looking forward to the changes in #5204 though, these changes are going to be more and more widespread.

I have one more v2 change coming down the line. In my opinion, the `Client` class won't be dramatically changed by it; there will be two new public methods, and one new private method.

> What happens if you call new_order on a client for ACMEv1 or request_challenges on a client for ACMEv2?

For new_order on an ACMEv1 client, we can pretty easily do the right thing: Turn it into a set of request_challenges. For request_challenges on ACMEv2 we could arguably synthesize an order out of them (now that you don't need a CSR to create an order), but it is probably better to just blow up.

In my opinion, on balance, the changes are not obtrusive enough to justify a whole new class.

Reply:
> For new_order on an ACMEv1 client, we can pretty easily do the right thing: Turn it into a set of request_challenges. For request_challenges on ACMEv2 we could arguably synthesize an order out of them (now that you don't need a CSR to create an order), but it is probably better to just blow up.

Do we want to do this work though? My main concern here is making it clear to both existing and future ACME clients what methods they can and should use for each ACME version. I think separate classes (with a common base class containing most of the functionality) makes this self documenting. If you're using an ACMEv1 client, you can't call `new_order` because the method doesn't exists. Another example that presumably doesn't make sense is calling `agree_to_tos` on an ACMEv2 client.

If we don't go with the separate classes approach, we can try to document in the docstring for the class or each method what you can and can't use for each ACME version, but this to me feels awkward and less clear with the only major downside to the separate class approach being some initial refactoring up front.

What are the benefits in your mind to keeping everything in the same class?